File: backend/app/api/public.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlalchemy.orm import Session
from typing import Optional

from app.core.database import get_db
from app.core.deps import get_current_active_user
from app.models.user import User
from app.models.generation import Generation
from app.models.public_share import PublicShare
from app.schemas.generation import GenerationIdAction, GenerationOut, GenerationPage, NodeTypeCount
from app.api.generations import _get_owned_generation, _to_out

router = APIRouter(prefix="/api/public", tags=["public"])


@router.post("", response_model=GenerationOut)
def share_generation(
    payload: GenerationIdAction,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """将当前用户自己的生成记录公开到画廊。幂等：重复公开直接返回。"""
    gen = _get_owned_generation(payload.generation_id, current_user, db)
    if not gen.public_share:
        db.add(PublicShare(user_id=current_user.id, generation_id=gen.id))
        db.commit()
        db.refresh(gen)
    return _to_out(gen, current_user, include_username=True)


@router.get("", response_model=GenerationPage)
def list_public_shares(
    keyword: Optional[str] = None,
    node_type: Optional[str] = None,
    skip: int = 0,
    limit: int = 20,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """公开画廊：所有用户公开的生成记录（可按 keyword / node_type 筛选，分页返回）。"""
    query = db.query(PublicShare)
    if keyword or node_type:
        query = query.join(Generation, PublicShare.generation_id == Generation.id)
    if keyword:
        query = query.filter(Generation.name.ilike(f"%{keyword}%"))
    if node_type:
        query = query.filter(Generation.node_type == node_type)
    total = query.count()
    page_limit = min(limit, 100)
    shares = (
        query.order_by(PublicShare.created_at.desc(), PublicShare.id.desc())
        .offset(skip)
        .limit(page_limit)
        .all()
    )
    # 类型筛选项：公开画廊全部记录按类型分组计数（不受 keyword/node_type 过滤影响）
    node_type_counts = [
        NodeTypeCount(node_type=row[0], count=row[1])
        for row in db.query(Generation.node_type, func.count(Generation.id))
        .join(PublicShare, PublicShare.generation_id == Generation.id)
        .group_by(Generation.node_type)
        .all()
    ]
    return GenerationPage(
        items=[
            _to_out(share.generation, current_user, include_username=True)
            for share in shares
            if share.generation
        ],
        total=total,
        skip=skip,
        limit=page_limit,
        node_type_counts=node_type_counts,
    )


@router.delete("/{generation_id}", response_model=GenerationOut)
def unshare_generation(
    generation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """将当前用户自己的生成记录从画廊撤下。"""
    gen = _get_owned_generation(generation_id, current_user, db)
    share = (
        db.query(PublicShare)
        .filter(PublicShare.generation_id == gen.id, PublicShare.user_id == current_user.id)
        .first()
    )
    if share:
        db.delete(share)
        db.commit()
        db.refresh(gen)
    return _to_out(gen, current_user, include_username=True)

File: backend-ts/src/api/public.ts
import { and, count, desc, eq, like } from 'drizzle-orm';
import type { FastifyInstance } from 'fastify';
import { getDb } from '../config/database.js';
import { generations, publicShares } from '../db/schema.js';
import { now } from '../shared/datetime.js';
import { toGenerationOut, type GenerationPage, type GenerationRow } from './generations.js';

/**
 * 公开画廊接口（对应 Python `app/api/public.py`）：
 * - POST /api/public（公开自己的生成记录到画廊，幂等）
 * - GET /api/public（画廊列表：所有用户公开记录，keyword / node_type / 分页）
 * - DELETE /api/public/:generation_id（从画廊撤下）
 */

export async function registerPublicRouter(app: FastifyInstance): Promise<void> {
  // 公开到画廊（幂等，仅限自己的记录）
  app.post(
    '/api/public',
    { preHandler: app.authenticate },
    async (request, reply) => {
      const genId = Number((request.body as { generation_id?: number } | undefined)?.generation_id);
      const db = getDb();
      const gen = db
        .select()
        .from(generations)
        .where(and(eq(generations.id, genId), eq(generations.userId, request.authUser!.id)))
        .get();
      if (!gen) return reply.code(404).send({ detail: '生成记录不存在' });
      const share = db.select().from(publicShares).where(eq(publicShares.generationId, gen.id)).get();
      if (!share) {
        db.insert(publicShares).values({ userId: request.authUser!.id, generationId: gen.id, createdAt: now() }).run();
      }
      return toGenerationOut(db, gen, request.authUser!.id, true);
    }
  );

  // 画廊列表（所有用户）
  app.get('/api/public', { preHandler: app.authenticate }, async (request) => {
    const q = (request.query ?? {}) as { keyword?: string; node_type?: string; skip?: string; limit?: string };
    const keyword = (q.keyword ?? '').trim();
    const nodeType = (q.node_type ?? '').trim();
    const skip = Number(q.skip ?? 0) || 0;
    const limit = Math.min(Number(q.limit ?? 20) || 20, 100);
    const db = getDb();
    const uid = request.authUser!.id;

    const filters = [eq(publicShares.generationId, generations.id)];
    if (keyword) filters.push(like(generations.name, `%${keyword}%`));
    if (nodeType) filters.push(eq(generations.nodeType, nodeType));
    const where = and(...filters);

    const total = db
      .select({ c: count() })
      .from(publicShares)
      .innerJoin(generations, eq(publicShares.generationId, generations.id))
      .where(where)
      .get()!.c;

    const rows = db
      .select({
        id: generations.id,
        userId: generations.userId,
        nodeType: generations.nodeType,
        name: generations.name,
        stageResults: generations.stageResults,
        resultUrl: generations.resultUrl,
        status: generations.status,
        createdAt: generations.createdAt,
      })
      .from(publicShares)
      .innerJoin(generations, eq(publicShares.generationId, generations.id))
      .where(where)
      .orderBy(desc(publicShares.createdAt), desc(publicShares.id))
      .limit(limit)
      .offset(skip)
      .all();

    const counts = db
      .select({ nodeType: generations.nodeType, c: count(generations.id) })
      .from(generations)
      .innerJoin(publicShares, eq(publicShares.generationId, generations.id))
      .groupBy(generations.nodeType)
      .all();

    return {
      items: rows.map((r) => toGenerationOut(db, r as GenerationRow, uid, true)),
      total,
      skip,
      limit,
      node_type_counts: counts.map((r) => ({ node_type: r.nodeType, count: r.c })),
    } satisfies GenerationPage;
  });

  // 从画廊撤下（仅限自己的记录）
  app.delete('/api/public/:generation_id', { preHandler: app.authenticate }, async (request, reply) => {
    const genId = Number((request.params as { generation_id: string }).generation_id);
    const db = getDb();
    const gen = db
      .select()
      .from(generations)
      .where(and(eq(generations.id, genId), eq(generations.userId, request.authUser!.id)))
      .get();
    if (!gen) return reply.code(404).send({ detail: '生成记录不存在' });
    db.delete(publicShares)
      .where(and(eq(publicShares.generationId, gen.id), eq(publicShares.userId, request.authUser!.id)))
      .run();
    return toGenerationOut(db, gen, request.authUser!.id, true);
  });
}
